File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/B-Минимум_на_отрезке/solution.cpp
#include <iostream>
#include <vector>
#include <deque>


using namespace std;


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(K)
    */
    int n, k;
    scanf("%d %d", &n, &k);
    vector<int> nums (n, 0);
    for (int i = 0; i < n; ++i) cin >> nums[i];
    deque<int> dq;
    vector<int> output;
    for (int idx = 0; idx < n; ++idx) {
        if (!dq.empty() && idx - dq.front() + 1 > k) dq.pop_front();
        while (!dq.empty() && nums[dq.back()] > nums[idx]) dq.pop_back();
        dq.push_back(idx);
        if (idx + 1 >= k) output.push_back(nums[dq.front()]);
    }
    for (size_t i = 0; i < output.size(); ++i) {
        if (i) cout << " ";
        cout << output[i];
    }
    if (!output.empty()) cout << "\n";
}

 
int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/B-Минимум_на_отрезке/solution.py
import sys
from collections import deque


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(K)
    """
    n, k = map(int, sys.stdin.readline().rstrip().split())
    nums: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    output: list[int] = []
    dq: deque[int] = deque()
    for idx in range(n):
        if dq and idx - dq[0] + 1 > k:
            dq.popleft()
        while dq and nums[dq[-1]] > nums[idx]:
            dq.pop()
        dq.append(idx)
        if idx >= k - 1:
            output.append(nums[dq[0]])
    sys.stdout.write(' '.join(str(min_num) for min_num in output))
    sys.stdout.write('\n')


if __name__ == '__main__':
    solution() 